### PR TITLE
fleetctl: set --driver=etcd if --endpoint is etcd-like

### PR DESCRIPTION
**Pull request by bcwaldon**
There is one remaining wart of the switch from --driver=etcd to --driver=API. If someone had been using fleetctl with a custom endpoint like so:

```
fleetctl --endpoint http://example.com:4001 list-machines
```

They now all of a sudden get directed to the fleet API rather than etcd. Given that we can tell (based on the port) that the user wants to use etcd, the default of --driver in this case should remain "etcd".

### CODE REVIEW

File: fleetctl/fleetctl.go [hunk 1]
```diff
@@ -262,6 +262,26 @@ func main() {
...
+
+	// if --driver is not set, but --endpoint looks like an etcd
+	// server, set the driver to etcd
+	if _, ok := visited["driver"]; !ok {
+		if _, ok = visited["endpoint"]; ok {
+			if eps := strings.Split(globalFlags.Endpoint, ","); len(eps) >= 1 {
+				if u, err := url.Parse(eps[0]); err == nil {
+					if _, port, err := net.SplitHostPort(u.Host); err == nil {
+						if port == "4001" || port == "2379" {
+							log.Debugf("Defaulting to --driver=%s as --endpoint appears to be etcd", clientDriverEtcd)
```
Comment (crawford) [inline fleetctl/fleetctl.go]: 7 levels of indentation?

Comment (bcwaldon) [inline fleetctl/fleetctl.go]: deal with it

Comment (jonboulle) [inline fleetctl/fleetctl.go]: UMM

Comment (crawford) [inline fleetctl/fleetctl.go]: How does this look if it's in an anonymous function, returning after failed conditions?

File: fleetctl/fleetctl.go [hunk 2]
```diff
@@ -262,6 +262,26 @@ func main() {
 		os.Exit(2)
 	}
 
+	visited := make(map[string]struct{}, 0)
+	globalFlagset.Visit(func(f *flag.Flag) { visited[f.Name] = struct{}{} })
+
+	// if --driver is not set, but --endpoint looks like an etcd
+	// server, set the driver to etcd
+	if _, ok := visited["driver"]; !ok {
```
Comment (crawford) [inline fleetctl/fleetctl.go]: `if  visited["driver"] {`

Comment (crawford) [inline fleetctl/fleetctl.go]: shit

File: fleetctl/fleetctl.go [hunk 3]
```diff
@@ -262,6 +262,26 @@ func main() {
...
+		if _, ok = visited["endpoint"]; ok {
```
Comment (crawford) [inline fleetctl/fleetctl.go]: `if visited["endpoint"] {`

File: fleetctl/fleetctl.go (hunk 1 above)
Comment (crawford) [inline fleetctl/fleetctl.go]: Or if you wrap the check in a recover() and just assume that the Split, Parse, and SplitHostPort worked.

Comment (crawford) [inline fleetctl/fleetctl.go]:
```go
if visited["driver"] == nil && visited["endpoint"] && u, err := url.Parse(strings.Split(globalFlags.Endpoint, ",")[0]); err == nil  {
 if _, port, err := net.SplitHostPort(u.Host); err == nil && (port == "4001" || port == "2379") {
  log.Debugf("Defaulting to --driver=%s as --endpoint appears to be etcd", clientDriverEtcd)
  globalFlags.ClientDriver = clientDriverEtcd
 }
}
```

**Comment by crawford** [comment]
LGTM

**Pull request merged**